File: backend/graph/tools/http_tool.py
"""
Generic HTTP tool: calls external APIs with credential injection.

Config schema (stored in tool_definitions.config JSONB):
{
    "url": "https://api.example.com/endpoint",
    "method": "POST",
    "headers_template": {"Authorization": "Bearer {{credential.api_key}}"},
    "body_template": {"message": "{{input.text}}"},
    "description": "Call the Example API"
}
"""

import logging
import re
from typing import Any, Optional

import httpx
from langchain_core.tools import StructuredTool

from app.services.vault import get_credential_decrypted

logger = logging.getLogger(__name__)


def _interpolate(template: Any, context: dict[str, Any]) -> Any:
    """Replace {{key.subkey}} placeholders in a template with context values."""
    if isinstance(template, str):
        def replace_match(match: re.Match) -> str:
            path = match.group(1).strip()
            parts = path.split(".")
            value = context
            for part in parts:
                if isinstance(value, dict):
                    value = value.get(part, "")
                else:
                    return ""
            return str(value) if value is not None else ""

        return re.sub(r"\{\{(.+?)\}\}", replace_match, template)

    elif isinstance(template, dict):
        return {k: _interpolate(v, context) for k, v in template.items()}
    elif isinstance(template, list):
        return [_interpolate(v, context) for v in template]
    return template


async def build_http_tool(
    name: str,
    config: dict[str, Any],
    credential_id: Optional[str] = None,
) -> StructuredTool:
    """
    Build a LangChain StructuredTool that makes an HTTP call.
    Credential is fetched and decrypted at call time.
    """
    url_template = config.get("url", "")
    method = config.get("method", "GET").upper()
    headers_template = config.get("headers_template", {})
    body_template = config.get("body_template", None)
    description = config.get("description", f"Call {name} HTTP API")

    async def _call_http(input_text: str = "") -> str:
        """Execute HTTP tool call with credential injection."""
        # Build interpolation context
        context: dict[str, Any] = {"input": {"text": input_text}}

        if credential_id:
            try:
                cred = await get_credential_decrypted(credential_id)
                context["credential"] = cred["config"]
            except Exception as e:
                return f"Error loading credential: {e}"

        url = _interpolate(url_template, context)
        headers = _interpolate(headers_template, context)
        body = _interpolate(body_template, context) if body_template else None

        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.request(
                    method=method,
                    url=url,
                    headers=headers,
                    json=body if body else None,
                )
                if response.status_code >= 400:
                    return f"HTTP {response.status_code}: {response.text[:500]}"
                return response.text[:2000]
        except Exception as e:
            return f"HTTP request failed: {type(e).__name__}: {str(e)}"

    return StructuredTool.from_function(
        coroutine=_call_http,
        name=name,
        description=description,
    )

File: frontend/worker/tools/http-tool.ts
/**
 * Generic HTTP tool: calls external APIs with credential injection.
 * Port of backend/graph/tools/http_tool.py.
 */

import { tool } from "ai";
import { z } from "zod";
import { getDb } from "../db";
import { getCredentialDecrypted } from "@/lib/vault";

function interpolate(template: unknown, context: Record<string, unknown>): unknown {
  if (typeof template === "string") {
    return template.replace(/\{\{(.+?)\}\}/g, (_, path: string) => {
      const parts = path.trim().split(".");
      let value: unknown = context;
      for (const part of parts) {
        if (value && typeof value === "object") {
          value = (value as Record<string, unknown>)[part];
        } else {
          return "";
        }
      }
      return value != null ? String(value) : "";
    });
  }
  if (Array.isArray(template)) return template.map((v) => interpolate(v, context));
  if (template && typeof template === "object") {
    const result: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(template)) {
      result[k] = interpolate(v, context);
    }
    return result;
  }
  return template;
}

export function buildHttpTool(
  name: string,
  config: Record<string, unknown>,
  credentialId: string | null,
) {
  const urlTemplate = (config.url as string) || "";
  const method = ((config.method as string) || "GET").toUpperCase();
  const headersTemplate = (config.headers_template as Record<string, string>) || {};
  const bodyTemplate = config.body_template || null;
  const description = (config.description as string) || `Call ${name} HTTP API`;

  return tool({
    description,
    inputSchema: z.object({
      input_text: z.string().default("").describe("Input text for the API call"),
    }),
    execute: async ({ input_text }: { input_text: string }) => {
      const context: Record<string, unknown> = { input: { text: input_text } };

      if (credentialId) {
        try {
          const db = getDb();
          const cred = await getCredentialDecrypted(db, credentialId);
          context.credential = cred.config;
        } catch (e) {
          return `Error loading credential: ${e}`;
        }
      }

      const url = interpolate(urlTemplate, context) as string;
      const headers = interpolate(headersTemplate, context) as Record<string, string>;
      const body = bodyTemplate ? interpolate(bodyTemplate, context) : null;

      try {
        const res = await fetch(url, {
          method,
          headers: { "Content-Type": "application/json", ...headers },
          body: body ? JSON.stringify(body) : undefined,
          signal: AbortSignal.timeout(30_000),
        });

        if (res.status >= 400) {
          const text = await res.text();
          return `HTTP ${res.status}: ${text.slice(0, 500)}`;
        }

        const text = await res.text();
        return text.slice(0, 2000);
      } catch (e) {
        return `HTTP request failed: ${e instanceof Error ? e.message : String(e)}`;
      }
    },
  });
}
